src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs

```diff
@@ -43,165 +32,111 @@ internal bool TryGetBuffer(in SequencePosition start, in SequencePosition end, o
-
-                data = (ReadOnlyMemory<T>)(object)(Unsafe.As<string>(startObject)).AsMemory();
-            }
-            else
-            {
-                data = default;
-                return false;
-            }
-
-            data = data.Slice(startIndex, length);
-            return true;
-        }
-
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        internal ReadOnlyMemory<T> GetFirstBuffer(in SequencePosition start, in SequencePosition end)
-        {
-            SequenceType type;
-            int startIndex = 0;
-            int endIndex = 0;
-            object startObject = start.GetObject();
-            if (startObject != null)
-            {
-                GetTypeAndIndices(start.GetInteger(), end.GetInteger(), out type, out startIndex, out endIndex);
-            }
             else
             {
-                type = SequenceType.Empty;
-            }
+                //if (startObject != end.GetObject())
+                //    ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
+                Debug.Assert(startObject == end.GetObject());
 
-            int length = endIndex - startIndex;
-            object endObject = end.GetObject();
-
-            if (type != SequenceType.MultiSegment && type != SequenceType.Empty && startObject != endObject)
-                ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
-
-            ReadOnlyMemory<T> memory;
-            if (type == SequenceType.MultiSegment)
-            {
-                Debug.Assert(startObject is ReadOnlySequenceSegment<T>);
-                ReadOnlySequenceSegment<T> startSegment = Unsafe.As<ReadOnlySequenceSegment<T>>(startObject);
-                memory = startSegment.Memory;
-                if (startObject != endObject)
+                if (type == SequenceType.Array)
                 {
-                    length = memory.Length - startIndex;
-                }
-            }
-            else if (type == SequenceType.Array)
-            {
-                Debug.Assert(startObject is T[]);
+                    Debug.Assert(startObject is T[]);
 
-                memory = new ReadOnlyMemory<T>(Unsafe.As<T[]>(startObject));
-            }
-            else if (type == SequenceType.OwnedMemory)
-            {
-                Debug.Assert(startObject is OwnedMemory<T>);
+                    data = new ReadOnlyMemory<T>(Unsafe.As<T[]>(startObject));
+                }
+                else if (type == SequenceType.OwnedMemory)
+                {
+                    Debug.Assert(startObject is OwnedMemory<T>);
 
-                memory = (Unsafe.As<OwnedMemory<T>>(startObject)).Memory;
-            }
-            else if (typeof(T) == typeof(char) && type == SequenceType.String)
-            {
-                Debug.Assert(startObject is string);
+                    data = (Unsafe.As<OwnedMemory<T>>(startObject)).Memory;
+                }
+                else
```

Comment:
Still want 
```csharp
else if (typeof(T) == typeof(char)
```
As if `T` is not `char` e.g. `byte` the Jit will then remove the block entirely